File: ceoi/2008/day2/snake/snakelib.pas
unit snakelib;

interface

function get_speed : longint;
procedure ask_snake(U1 : longint; U2 : longint; var A1 : char; var A2 : char);
procedure tell_length(L : longint);

implementation

const MAXN = 12121;
   MAXLEN = 12122;
   MAXCALL1 = 13;
   MAXCALL2 = 42;
   MAXSPEED = 10;
   
INPUT_FILE_NAME = 'snake.in';
   LOG_FILE_NAME = 'snake.log';

var initialised : boolean;
   infile, logfile : text;

Len, Tail, K : longint;
   turn : longint;

procedure die(Message : string);
var ior : longint;

begin
   if not initialised then
   begin
      assign(logfile, LOG_FILE_NAME);
      {$I-} rewrite(logfile); {$I+}
      ior := ioresult;
      if ior <> 0 then 
      begin
	 writeln('Error opening snake.log for writing.');
	 halt(1);
      end;
   end;
   writeln('An error has occured: ', Message, '. Please check snake.log');
   writeln(logfile, 'Error ', Message, '!');
   close(logfile);
   halt(1);
end;

function inttostr(i : longint) : string;
var s : string;
begin
   str(i, s);
   inttostr := s;
end;

procedure initialise;
var tmp : string;
   ior : longint;
begin
   assign(logfile, LOG_FILE_NAME);
   {$I-} rewrite(logfile); {$I+}
   ior := ioresult;
   if ior <> 0 then 
   begin
      writeln('Error opening snake.log for writing.');
      halt(1);
   end;

   assign(infile, INPUT_FILE_NAME);
   {$I-} reset(infile); {$I+}
   ior := ioresult;
   if ior <> 0 then die('could not open snake.in');
   read(infile, Len, Tail, K);
   close(infile);
   if Len > MAXLEN then begin tmp := 'the snake length must be at most ' + inttostr(MAXLEN); die(tmp); end;
   if Len < 1 then die('the snake must have a positive length');
   if Tail + Len - 1 > MAXN then begin tmp := 'all parts of the snake must initially be within the hearing distance ' + inttostr(MAXN); die(tmp); end;
   if Tail < 0 then die('the initial distance between you and the snake cannot be negative');
   if K < 0 then die('the snake speed cannot be a negative number');
   if K > MAXSPEED then begin tmp := 'the snake speed cannot exceed ' + inttostr(MAXSPEED); die(tmp); end;

   initialised := true;
end;


function get_speed : longint;
begin
   if initialised then die('do not call get_speed more than once');
   if not initialised then initialise;
   writeln(logfile, 'get_speed. reply: ', inttostr(K));
   get_speed := K;
end;

function eval(p : longint) : char;
var a : char;
   tmp : string;
begin
   if p < 0 then die('the position cannot have a negative value');
   if p > MAXN then begin tmp := 'the position cannot exceed ' + inttostr(MAXN); die(tmp); end;
   if p < Tail then begin a := 'b' end
   else if (p >= Tail + Len) then begin a := 'f' end
   else begin a := 's'; end;
   eval := a;
end;

procedure ask_snake(U1 : longint; U2 : longint; var A1 : char; var A2 : char);
var tmp : string;
begin
   inc(turn);
   if not initialised then die('you need to call get_speed first');
   write(logfile, turn:2, ': ask_snake(', U1, ', ', U2, '). reply: ');
   if turn >= MAXCALL2 then begin tmp := 'you cannot ask the snake more than ' + inttostr(MAXCALL2 - 1) + ' times'; die(tmp); end;

   if (turn mod 2 = 0) AND (Tail + K + Len < MAXN) then inc(Tail, K);
   A1 := eval(U1);
   A2 := eval(U2);
   writeln(logfile, A1, ' ', A2);
   writeln(logfile, '    (the snake currently covers units ', Tail, ' - ', Tail + Len - 1, ')');
end;

procedure tell_length(L : longint);
begin
   if not initialised then die('you need to call get_speed first');
   writeln(logfile, 'tell_length(', L, ')');
   if L < 1 then die('the snake length is always a positive number');
   if abs(L - Len) > K then 
      writeln(logfile, 'Wrong! Your estimate is not good enough')
   else
   begin
      writeln(logfile, 'Your estimate is within the allowed range! You needed to ask the snake ', turn, ' times.');
      if (turn <= MAXCALL1) then 
	 writeln(logfile, 'You will get full score for this test')
      else 
	 writeln(logfile, 'You will get half score for this test');
   end;
   close(logfile); 
   halt(0);
end;

begin
   initialised := false;
   turn := 0;
end.

File: ceoi/2008/day2/snake/snakelib.c
#include <stdio.h>
#include <stdlib.h>

#include "snakelib.h"

#define INPUT_FILE_NAME "snake.in"
#define LOG_FILE_NAME "snake.log"

static int initialised = 0;
static int Len, Tail, K, turn = 0;
static FILE *log = NULL;

static void die(char const* Message) {
   printf("An error occured: %s. Please check snake.log!\n", Message);
   if (log != NULL) {
      fprintf(log, "Error: %s!\n", Message);
      fclose(log);
   } else {
      log = fopen(LOG_FILE_NAME, "w");
      if (log == NULL) {
	 printf("error opening snake.log for writing");
	 exit(1);
      }
      fprintf(log, "Error: %s!\n", Message);
      fclose(log);
   }
   exit(1);
}

void initialise()
{
   FILE *in;
   char tmp[256];
  
   log = fopen(LOG_FILE_NAME, "w");
   if (log == NULL) {
      printf("error opening snake.log for writing");
      exit(1);
   }
  
   in = fopen(INPUT_FILE_NAME, "r");
   if (in == NULL)
      die("could not open snake.in");
  
   if (fscanf(in, "%d %d %d", &Len, &Tail, &K) != 3) 
      die("Wrong format in snake.in");
   if (Len > MAXLEN) {
      sprintf(tmp, "the snake length must be at most %d", MAXLEN);
      die(tmp);
   }
   if (Len < 1) die("the snake must have a positive length");
   if (Tail + Len - 1 > MAXN) {
      sprintf(tmp, "all parts of the snake must initially be within the hearing distance %d", MAXN);
      die(tmp);
   }
   if (Tail < 0) die("the initial distance between you and the snake cannot be negative");
   if (K < 0) die("the snake speed cannot be a negative number");
   if (K > MAXSPEED) {
      sprintf(tmp, "the snake speed cannot exceed %d", MAXSPEED);
      die(tmp);
   }

   initialised = 1;
}

int get_speed()
{
   if (initialised) die("do not call get_speed more than once");
   if (!initialised) initialise();
   fprintf(log, "get_speed. reply: %d\n", K);
   return K;
}

char eval(int p)
{
   char a;
   char tmp[256];
  
 
   if (p < 0) {
      die("the position cannot have a negative value");
   }
   if (p > MAXN) {
      sprintf(tmp, "the position cannot exceed %d", MAXN);
      die(tmp);
   }
   if (p < Tail) {
      a = 'b';
   } else if (p >= Tail + Len) {
      a = 'f';
   } else {
      a = 's';
   }
   return a;
}

void ask_snake(int U1, int U2, char *A1, char *A2)
{
   if (!initialised)
      die("you need to call get_speed first");
  
   turn++;
   char tmp[256];
   fprintf(log, "%2d: ask_snake(%d, %d). reply: ", turn, U1, U2);
   if (turn >= MAXCALL2) { sprintf(tmp, "you cannot ask the snake more than %d times", MAXCALL2 - 1); die(tmp); }
   if ((turn % 2 == 0) && Tail + K + Len <= MAXN) Tail += K;
   *A1 = eval(U1);
   *A2 = eval(U2);
   fprintf(log, "%c %c\n", *A1, *A2);
   fprintf(log, "    (the snake currently covers units %d - %d)\n", Tail, Tail + Len - 1);
}

void tell_length(int L) {
   if (!initialised) die("you need to call get_speed first");
   fprintf(log, "tell_length(%d)\n", L);
   if (L < 1) die("the snake length is always a positive number");
   if (abs(L - Len) > K) {
      fprintf(log, "Wrong! Your estimate is not good enough");
   } else {
      fprintf(log, "Your estimate is within the allowed range! You needed to ask the snake %d times\n", turn);
      fprintf(log, "You would get %s score for this test\n", turn <= MAXCALL1 ? "full" : "half");
   }
   fclose(log);
   exit(0);
}
